### PR TITLE
Spike: Constrain the OpenAPI schema emitted for a model [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by bajtos**
This pull request offers a PoC implementation demonstrating how to modify the schema generated from a LB4 model depending on configurable options. In particular, the following two use cases are covered:

- PATCH - all properties are optional
- CREATE - id and possibly _rev must not be present

The first commit imports code from https://github.com/strongloop/loopback-next/pull/2592 and should be ignored.

The second commit contains PoC and SPIKE notes with a high-level overview.

Related stories:
 - the spike story: https://github.com/strongloop/loopback-next/issues/2082
 - #1722
 - #1179
 - #1470
 - #1719
 - #1894

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: packages/repository-json-schema/src/build-schema.ts [hunk 1]
```diff
@@ -14,8 +14,16 @@ import {
 import {JSONSchema6 as JSONSchema} from 'json-schema';
 import {JSON_SCHEMA_KEY} from './keys';
 
-export interface JsonSchemaOptions {
+// tslint:disable-next-line:no-any
+export interface JsonSchemaOptions<T extends object = any> {
   includeRelations?: boolean;
+
+  /// List of properties to exclude from the schema
+  exclude?: (keyof T)[];
```
Comment (bajtos) [inline packages/repository-json-schema/src/build-schema.ts]: This type construct allows TypeScript to verify that `exclude` is using only existing model properties.

Consider the following incorrect usage:

```ts
{
  schema: getModelSchemaRef(TodoList, {exclude: ['id', 'foobar']}),
}
```

The compiler prints the following error:

```
Argument of type '{ exclude: ("id" | "foobar")[]; }' is not assignable to parameter of type 'JsonSchemaOptions<TodoList>'.
  Types of property 'exclude' are incompatible.
    Type '("id" | "foobar")[]' is not assignable to type '("id" | "title" | "color" | "todos" | "image" | "getId" | "getIdObject" | "toJSON" | "toObject")[]'.
      Type '"id" | "foobar"' is not assignable to type '"id" | "title" | "color" | "todos" | "image" | "getId" | "getIdObject" | "toJSON" | "toObject"'.
        Type '"foobar"' is not assignable to type '"id" | "title" | "color" | "todos" | "image" | "getId" | "getIdObject" | "toJSON" | "toObject"'.
```

File: _SPIKE_.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  async create(
+    @requestBody({
+      content: {
+        'application/json': {
+          schema: getModelSchemaRef(TodoList, {exclude: ['id']}),
+          /***** ^^^ THIS IS IMPORTANT - OPENAPI SCHEMA ^^^ ****/
+        },
+      },
+    })
+    obj: Pick<TodoList, Exclude<keyof TodoList, 'id'>>,
```
Comment (raymondfeng) [inline _SPIKE_.md]: Maybe define a type alias:
```ts
export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
```

Comment (bajtos) [reply _SPIKE_.md]: +1 to define a type alias.

We must carefully choose a name that will be different enough from `Exclude`. Few options that come to my mind:
- `ExcludeProperties<T, K>`
- `ExcludeKeys<T, K>`
- `OmitProperties<T, K>`
- `OmitKeys<T, K>`

**Comment by bajtos** [comment]
Created the following stories:
- Emit schema with all model properties optional #2652
- Emit schema excluding certain model properties #2653
- Spike: simplify requestBody annotation with schema options #2654

Closing this spike as done.

**Pull request closed without merging**